### DISCUSSION TITLE
Ideas & Concerns

### DISCUSSION THREAD

**Discussion by Tired-Fox** [Tired-Fox/mlua-extras#1] [Q&A]
This project is early and experimental. Please share any ideas or concerns here.

**Comment by ian-h-chamberlain** [comment]
Hi, not sure if this is still under development but it looks like it a very interesting project! I wanted to try it out but I can't seem to get the example to compile on `main` or `develop` branches...

```console
$ cargo check --features vendored
   Compiling mlua-sys v0.6.3
error: You can enable only one of the features: lua54, lua53, lua52, lua51, luajit, luajit52, luau
  --> /Users/ianchamberlain/.config/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/mlua-sys-0.6.3/build/main.rs:16:13
   |
16 |             compile_error!("You can enable only one of the features: lua54, lua53, lua52, lua51, luajit, luajit52, luau");
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `mlua-sys` (build script) due to 1 previous error
```

Any idea what I'm missing here? Is there some other combination of feature flags I should be using? I'll try to set up the right environment for `lua54` or something meanwhile but just asking to make sure I'm not missing something obvious.

**Reply by ian-h-chamberlain** [reply]
Update: `lua54` seems to be working after setting up my environemnt, although I'm not sure why `vendored` didn't.


Now I'm running into another set of errors while trying to use the `send` feature; it appears to fail on all the helper wrapper methods for a missing `Send` bound, which seems odd because I was able to compile the examples just fine! I don't even have to add any code before trying this to see it fail:

```
error[E0277]: the trait bound `SET: mlua::types::MaybeSend` is not satisfied
   --> /Users/ianchamberlain/Documents/Development/mlua-extras/src/extras/mod.rs:399:43
    |
399 |         self.add_field_function_set(name, set);
    |              ----------------------       ^^^ the trait `Send` is not implemented for `SET`
    |              |
    |              required by a bound introduced by this call
    |
    = note: required for `SET` to implement `mlua::types::MaybeSend`
note: required by a bound in `add_field_function_set`
   --> /Users/ianchamberlain/.config/cargo/registry/src/index.crates.io-1949cf8c6b5b557f/mlua-0.9.9/src/userdata.rs:534:67
    |
532 |     fn add_field_function_set<F, A>(&mut self, name: impl AsRef<str>, function: F)
    |        ---------------------- required by a bound in this associated function
533 |     where
534 |         F: FnMut(&'lua Lua, AnyUserData<'lua>, A) -> Result<()> + MaybeSend + 'static,
    |                                                                   ^^^^^^^^^ required by this bound in `UserDataFields::add_field_function_set`
help: use parentheses to call this type parameter
    |
399 |         self.add_field_function_set(name, set(/* &'lua Lua */, /* AnyUserData<'lua> */, /* A */));
    |                                              +++++++++++++++++++++++++++++++++++++++++++++++++++
help: consider further restricting type parameter `SET` with trait `Send`
    |
396 |         SET: 'static + MaybeSend + Fn(&'lua Lua, AnyUserData<'lua>, A) -> mlua::Result<()> + std::marker::Send,
    |                                                                                            +++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `mlua-extras` (lib) due to 35 previous errors
```

Any ideas about what might be going on here or if there's a way to resolve it?

**Reply by Tired-Fox** [reply]
@ian-h-chamberlain

For the vendored feature flag, it follows `mlua`'s feature flag scheme so you will have to specify a lua version like so `["vendored", "lua54"]`.

For the error above, it looks like the `mlua` `"send"` feature is enabled. This will make it so that the `MaybeSend` trait will require anything that implements it to also implement `Send`.  The fix for this would be to either remove the `"send"` feature from `mlua` or add the `"send"` feature to this crate in your `Cargo.toml`.

**Comment by StayBlue** [comment]
Hello! I stumbled upon this crate, and the automatic generation of documentation is something that'd be really useful for my project. Will the mlua version be updated?

**Reply by Tired-Fox** [reply]
@StayBlue 

Yes. I can look at bumping the version this week if you are still interested
